onnx/src/ops/ml/mod.rs

```diff
@@ -0,0 +1,330 @@
+
+    let inc_by_1 = |x: &[_]| x.iter().zip(x.iter().skip(1)).all(|(&x, &y)| y == x || y == x + 1);
+    node.expect_attr("nodes_treeids", inc_by_1(&tree_ids), "tree ids to increase by 1")?;
+    node.expect_attr(&cls("treeids"), inc_by_1(&leaf_tree_ids), "leaf tree ids to increase by 1")?;
+    node.expect_attr("nodes_treeids", tree_ids[0] == 0, "tree ids to start from 0")?;
+    node.expect_attr(&cls("treeids"), leaf_tree_ids[0] == 0, "leaf tree ids to start from 0")?;
+    let n_trees = *tree_ids.last().unwrap();
+    node.expect(leaf_tree_ids.last() == Some(&n_trees), "mismatching # of trees (nodes/leaves)")?;
+
+    // build the leaf map and collect all leaves
+
+    let mut leaf_map: HashMap<(usize, usize), (usize, usize)> = HashMap::default();
+    let mut leaves: Vec<Vec<Leaf>> = iter::repeat_with(Vec::default).take(n_trees).collect();
+
+    let mut prev_leaf_tree_id = -1i64 as usize;
+    let mut prev_leaf_node_id = 0;
+
+    for i in 0..n_leaves {
+        let leaf_tree_id = leaf_tree_ids[i];
+        let leaf_node_id = leaf_node_ids[i];
+        let leaf_class_id = leaf_class_ids[i];
+        let leaf_weight = leaf_weights[i];
+
+        if leaf_tree_id == prev_leaf_tree_id {
+            node.expect(leaf_node_id >= prev_leaf_node_id, "leaf node ids must be increasing")?;
+        }
+        let tree_leaves = &mut leaves[leaf_tree_id];
+        leaf_map
+            .entry((leaf_tree_id, leaf_node_id))
+            .or_insert_with(|| (tree_leaves.len(), tree_leaves.len()))
+            .1 += 1;
+        tree_leaves.push(Leaf::new(leaf_class_id, leaf_weight));
+
+        prev_leaf_tree_id = leaf_tree_id;
+        prev_leaf_node_id = leaf_node_id;
+    }
+
+    // collect all nodes
+
+    let mut nodes: Vec<Vec<Node>> = Vec::default();
+
+    let mut prev_tree_id = -1i64 as usize;
+    let mut prev_node_id = 0;
+
+    for i in 0..n_nodes {
+        let tree_id = tree_ids[i];
+        let node_id = node_ids[i];
+        let node_mode = node_modes[i];
+
+        if tree_id != prev_tree_id {
+            node.expect(node_id == 0, "node ids must start from 0 for each tree")?;
+        } else {
+            node.expect(node_id == prev_node_id + 1, "node ids must increase by 1")?;
+        }
+        let tree_node = if let Some(cmp) = node_mode {
+            let feature_id = feature_ids[i];
+            let value = node_values[i];
+            let true_id = true_ids[i];
+            let false_id = false_ids[i];
+            let nan_is_true = nan_is_true[i];
+            Node::new_branch(cmp, feature_id, value, true_id, false_id, nan_is_true)
+        } else {
+            if let Some(&(start, end)) = leaf_map.get(&(tree_id, node_id)) {
+                Node::new_leaf(start, end)
+            } else {
+                node.bail(&format!("leaf not found: tree_id={}, node_id={}", tree_id, node_id))?
+            }
+        };
+        nodes[tree_id].push(tree_node);
+
+        prev_tree_id = tree_id;
+        prev_node_id = node_id;
+    }
```

Comment:
about this part, which is indeed a bit tedious, have you consider using the zip macro from itertools to walk all the vectors in sync once your done with the sanity checks ?

Reply:
I think this part is actually overly tedious, the worst one is constructing the leaf map above... and in all cases I kind of need to know the node_id/tree_id on the previous step (well, sort of... it could be done in a different way, but it would be much harder to read). I think with itertools it would end up much harder to read anyway if you were to zip it between each other and also with previous steps.